core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -0,0 +1,379 @@
+package org.infinispan.persistence.spi;
+
+import java.util.EnumSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Predicate;
+
+import javax.transaction.Transaction;
+
+import org.infinispan.commons.dataconversion.MediaType;
+import org.infinispan.commons.util.Experimental;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.persistence.support.BatchModification;
+import org.infinispan.util.concurrent.CompletableFutures;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+
+import io.reactivex.rxjava3.core.Completable;
+import io.reactivex.rxjava3.core.Flowable;
+
+/**
+ * <p>
+ * Implementations of this store must be thread safe if concurrent operations are performed on it. This should include
+ * possibly invoking start or stop multiple times
+ * <p>
+ * Note that this method is Experimental and its methods may change slightly over time until it has matured.
+ * @author William Burns
+ * @since 11.0
+ * @param <K> key value type
+ * @param <V> value value type
+ */
+@Experimental
+public interface NonBlockingStore<K, V> {
+
+   enum Characteristic {
```

Comment:
what is your idea behind this? 
It doesn't seem very useful to me IMO (probably only `SHAREABLE` would make sense).
If a store is "write-only", return null on `load()`. A store doesn't support bulk reads, then return an empty publisher? No segments? ignore the parameter...

Reply:
We have logic around whether a Store can only write or read. For example if an entry is expired, we only read from Stores that can't also write.

Bulk reads we have to be able to detect, because we can support more than 1 store, we only ask the first store that supports the bulk read, so we have to be able to tell which one is which.

Segmentation is also special, as we have to know if a store was able to remove its segments during state transfer or not. If not then we have to remove the values manually, which is much slower.

The only option I would have to really deep down reevaluate is the WRITE_ONLY as I don't know if we have code that specifically cares about that. I might be able to remove that.

Reply:
> Segmentation is also special, as we have to know if a store was able to remove its segments during state transfer or not. If not then we have to remove the values manually, which is much slower.
>

I disagree. The store impl has access to the `KeyPartitioner` so, we can "force" the store to implement `removeSegments()` and suggest to use the `KeyPartitioner` to find the segment of the key (if needed).

The bulk read seems a big thing. Without it, you don't have state transfer. Not sure if a store is useful without it.

`SHAREABLE` can also be removed. It is a configuration parameter and, for example, the JDBC is shareable but you may want to have a single instance per node...

Reply:
> > Segmentation is also special, as we have to know if a store was able to remove its segments during state transfer or not. If not then we have to remove the values manually, which is much slower.
> 
> I disagree. The store impl has access to the `KeyPartitioner` so, we can "force" the store to implement `removeSegments()` and suggest to use the `KeyPartitioner` to find the segment of the key (if needed).

I agree, this is a possibility, however it also requires the implementer to do this. The point is also to try to lower the barrier for external users. We already support removing these individual entries across all stores in State transfer if the store implements `publishKeys`.

> The bulk read seems a big thing. Without it, you don't have state transfer. Not sure if a store is useful without it.

It is still quite useful to have a store that has a load but not a bulk read. For example many custom stores do this to read from the database etc. Many times this type of store is also  READ_ONLY.

> `SHAREABLE` can also be removed. It is a configuration parameter and, for example, the JDBC is shareable but you may want to have a single instance per node...

The point is the opposite, so you can't configure a store as SHARED that isn't SHAREABLE. For example you don't want to share a file based or LevelDB based store between nodes. I guess we can remove this, but it was sort of a way to verify that a store was configured properly. I personally would say to keep it unless you have a real preference against it.

Reply:
ok :+1:

Reply:
> I personally would say to keep it unless you have a real preference against it.

:+1: